woocommerce/google-listings-and-ads
# Create associated objects for Ads Campaigns

### Changes proposed in this Pull Request:

Closes #355 and might also close #332?

New functionality + [refactorathon](https://user-images.githubusercontent.com/228780/112556387-60b02980-8dca-11eb-832e-3f47454e4293.gif).

- Breaks out the Ads API features into their own classes (working with traits was too messy as almost all have dependencies that would have to come from the _using_ class, like the GoogleAdsClient or the Ads ID):
    - `AdsCampaign`
    - `AdsCampaignBudget`
    - `AdsGroup` (NOTE: this currently handles AdGroup, AdGroupAd, and Product Listing)
    - `AdsConversionAction`
- Creates the new `AdsIdTrait` and `AdsQueryTrait` that are shared among all the classes. Unsure where the `parse_id` method should live.
- In `AdsGroup`, one `public` method is provided to create all the objects needed for the campaign, using the specified default parameters. In the future if there will be multiple Product Listings/Groups, then that will probably need to be spun off.

**Notes**
- All the `Ads` classes have the same dependencies, may a `GoogleAdsClientAwareInterface` would make sense. And/or using the setter for the `$ads_id` in the "child" classes (like CampaignBudget and Group) instead of always providing it as an argument (it's already updated for `AdsConversionAction` because that's the only class that's used _during_ Ads account setup).
    https://github.com/woocommerce/google-listings-and-ads/blob/012530a051ed21fbd8d91f57329e5635b570c93d/src/Internal/DependencyManagement/GoogleServiceProvider.php#L89-L100
- Should the default campaign name "Ads Campaign [date]" be localized (so it matches the user's site/Ads dashboard language)?
    https://github.com/woocommerce/google-listings-and-ads/blob/b8cfa437b0b8bdf7626c519546816d6385a7a5eb/js/src/setup-ads/ads-stepper/setup-billing/index.js#L40-L46
    ![image](https://user-images.githubusercontent.com/228780/112557154-1465e900-8dcc-11eb-880b-3c5d67019ced.png)
- The ad group name is just `[CAMPAIGN NAME] Ad Group`. Not localized, nor is it updated if the campaign name is changed (something to consider in a future upgrade).
- A bit more controlled exception handling may be necessary (instead of just tossing in `@throws` everywhere).


### Screenshots:

_The created objects: Campaign, Ad Group, Ad Group Ad, and Product group ("Listing")_
![screenshot](https://d.pr/i/thHtPr+)


### Detailed test instructions:

1. Use the `Add Paid Campaign` button on the dashboard: `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard` to initiate the Ads setup flow.
2. After connecting an Ads account:
    - Navigate to any store page 
    - Search the source code for `gtag` 
    - Confirm the tracking code has been retrieved and inserted with the snippet (`AW-XXXXX`)
3. After setting up the paid campaign (using an existing sub-account seems to allow skipping the billing step):
   - Confirm all the objects existing the Ads panel: <https://ads.google.com/aw/overview> (see screenshot above)
4. Check that the campaign is retrieved correctly:
     - Connection Test `Get Campaigns from Google Ads` button
     - API `GET /wp-json/wc/gla/ads/campaigns`
     - Also, API `GET /wp-json/wc/gla/ads/campaigns/[ID]`
5. Update the campaign and check that the changes are made
     - API `POST /wp-json/wc/gla/ads/campaigns/[ID]` and
    ```json
    {
        "name": "[CHANGED NAME - emojis work 🙂 ]",
        "amount": 0.01,
        "country": "[??]"
    }
    ```
    - Confirm changes with API and in the Ads panel
 6. Delete the campaign
    - API `DELETE /wp-json/wc/gla/ads/campaigns/[ID]`
    - Confirm changes with API and in the Ads panel (worth noting that `GET .../campaigns/[ID]` will still retrieve the campaign with `"status": "removed"`, but the generic list `GET` won't.

**BONUS**
Check that the configuration for the ad group, ad group ad, and product listing all seem correct (in `AdsGroup`)
